service/frontend/health_check.go

```diff
@@ -91,17 +95,49 @@ func (h *healthCheckerImpl) Check(ctx context.Context) (enumsspb.HealthState, er
 	}
 
 	var failedHostCount float64
+	var hostDeclinedServingCount float64
 	for i := 0; i < len(hosts); i++ {
 		healthState := <-receiveCh
-		if healthState == enumsspb.HEALTH_STATE_NOT_SERVING {
+		switch healthState {
+		case enumsspb.HEALTH_STATE_NOT_SERVING, enumsspb.HEALTH_STATE_UNSPECIFIED:
 			failedHostCount++
+		case enumsspb.HEALTH_STATE_DECLINED_SERVING:
+			hostDeclinedServingCount++
+		case enumsspb.HEALTH_STATE_SERVING:
+			// Do nothing.
 		}
 	}
 	close(receiveCh)
 
-	if (failedHostCount / float64(len(hosts))) > h.hostFailurePercentage() {
+	// Make sure that at lease 2 hosts must be not ready to trigger this check.
+	proportionOfDeclinedServiceHosts := getProportionOfDeclinedServiceHosts(hostDeclinedServingCount/float64(len(hosts)), len(hosts))
+
+	hostDeclinedServingProportion := hostDeclinedServingCount / float64(len(hosts))
+	if hostDeclinedServingProportion > proportionOfDeclinedServiceHosts {
+		h.logger.Warn("health check exceeded host declined serving proportion threshold", tag.NewFloat64("host declined serving proportion threshold", proportionOfDeclinedServiceHosts))
+		return enumsspb.HEALTH_STATE_DECLINED_SERVING, nil
+	}
+
+	failedHostCountProportion := failedHostCount / float64(len(hosts))
+	if failedHostCountProportion+hostDeclinedServingProportion > h.hostFailurePercentage() {
+		h.logger.Warn("health check exceeded host failure percentage threshold", tag.NewFloat64("host failure percentage threshold", h.hostFailurePercentage()), tag.NewFloat64("host failure percentage", failedHostCountProportion), tag.NewFloat64("host declined serving percentage", hostDeclinedServingProportion))
 		return enumsspb.HEALTH_STATE_NOT_SERVING, nil
 	}
 
 	return enumsspb.HEALTH_STATE_SERVING, nil
 }
+
+func getProportionOfDeclinedServiceHosts(proportionOfDeclinedServingHosts float64, totalHosts int) float64 {
+	var numHostsToFail float64
+	if proportionOfDeclinedServingHosts > 0.0 { // sanity guard against panic.
+		numHostsToFail = float64(totalHosts) * proportionOfDeclinedServingHosts
+	}
+	var proportionOfDeclinedServiceHosts float64
+	if numHostsToFail < 2 {
+		correctedProportion := 2.0 / float64(totalHosts)
+		proportionOfDeclinedServiceHosts = math.Max(proportionOfDeclinedServingHosts, correctedProportion)
+	} else {
+		proportionOfDeclinedServiceHosts = proportionOfDeclinedServingHosts
+	}
+	return proportionOfDeclinedServiceHosts
```

Comment:
`if 1 / proportionOfDeclinedServingHosts > totalHost {return 1/ totalHosts} else {return proportionOfDeclinedServingHosts}`

Reply:
I've done you one better :) I definitely overcomplicated the first iteration

Reply:
Yes. This is much better.